### ISSUE TITLE
Build issues with 1.6.2 on older Debian releases, on i386, and on ppc64el

### ISSUE THREAD

**Issue by df7cb** [tada/pljava#338]
1.6.2 is working fine on the newer Debian/Ubuntu releases on amd64 and arm64, but unfortunately that doesn't cover everything on apt.postgresql.org:
https://pgdgbuild.dus.dg-i.net/job/postgresql-pljava-binaries/91/

## Failure on Debian stretch and Ubuntu xenial (all archs) (failing on PG9.5, and the others are not tried anymore):
https://pgdgbuild.dus.dg-i.net/job/postgresql-pljava-binaries/91/architecture=amd64,distribution=stretch/console
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project pljava-api: Fatal error compiling: invalid flag: --release -> [Help 1]
```
## Failure on all i386 distributions, all PG versions:
https://pgdgbuild.dus.dg-i.net/job/postgresql-pljava-binaries/91/architecture=i386,distribution=buster/console
```
psql -c CREATE EXTENSION pljava
ERROR:  java.sql.SQLException: An attempt was made to call a PostgreSQL backend function after an elog(ERROR) had been issued
```

## ppc64el, Debian buster + Ubuntu focal, PostgreSQL 12 (only):
https://pgdgbuild.dus.dg-i.net/job/postgresql-pljava-binaries/91/architecture=ppc64el,distribution=buster/consoleFull#console-section-18
https://pgdgbuild.dus.dg-i.net/job/postgresql-pljava-binaries/91/architecture=ppc64el,distribution=focal/consoleFull#console-section-18
```
21:06:48 # A fatal error has been detected by the Java Runtime Environment:
21:06:48 #
21:06:48 #  SIGSEGV (0xb) at pc=0x00007fff48011138, pid=10817, tid=10817
21:06:48 #
21:06:48 # JRE version: OpenJDK Runtime Environment (11.0.9.1+1) (build 11.0.9.1+1-post-Debian-1deb10u2)
21:06:48 # Java VM: OpenJDK 64-Bit Server VM (11.0.9.1+1-post-Debian-1deb10u2, mixed mode, tiered, compressed oops, g1 gc, linux-ppc64le)
21:06:48 # Problematic frame:
21:06:48 # C  0x00007fff48011138
```
(See full log for full trace.)

Interestingly, it does not fail at build-time where the test code is executed with explicit paths to .jar files and a (wide open) custom policy file. The crash happens at installed-time only.

Both these worked when the build was retried, though: https://pgdgbuild.dus.dg-i.net/job/postgresql-pljava-binaries/92/

**Comment by df7cb** [comment]
(Ignore the Jenkins runs marked as ABORTED, that's local networking weirdness.)

**Comment by jcflack** [comment]
The stretch/xenial issue (unrecognized `--release` option) is a consequence of trying to build with (it looks like) Java 8. PL/Java 1.6.x is strictly for Java 9 and later.

If a platform has Java 9 or later available but not as default, it will have to be treated as a dependency. It is not necessary to have the same version at build time and run time, but both must be 9 or later. For run time, either a Java 9 or later runtime can be installed as the box's default, or (if some other application requirement precludes that) it can simply be installed as non-default, and `pljava.libjvm_location` adjusted to point to it.

If a platform has no Java 9 or later JDK available at all, that platform will have to stay with PL/Java 1.5.x.

**Comment by jcflack** [comment]
Could the `i386` tests be run with `SET log_min_messages TO debug1` ?

The `ppc` segfault appears to be happening in LLVM-generated code again, like #277, which gives me a bad feeling. If it is stable with `SET jit_expressions TO off` or with `SET jit TO off`, that might have to be a release note for this platform, until I have more time to spend chasing that cause.

**Comment by df7cb** [comment]
I'll disable the older distributions since they have java 8 only, thanks.

For i386, there's nothing interesting even with debug5 in the server log:
```
2021-02-05 14:26:26.014 CET [462655] DEBUG:  forked new backend, pid=462715 socket=11
2021-02-05 14:26:26.014 CET [462715] cbe@postgres DEBUG:  postgres child[462715]: starting with (
2021-02-05 14:26:26.015 CET [462715] cbe@postgres DEBUG:  	postgres
2021-02-05 14:26:26.015 CET [462715] cbe@postgres DEBUG:  )
2021-02-05 14:26:26.015 CET [462715] cbe@postgres DEBUG:  InitPostgres
2021-02-05 14:26:26.015 CET [462715] cbe@postgres DEBUG:  my backend ID is 3
2021-02-05 14:26:26.015 CET [462715] cbe@postgres DEBUG:  StartTransaction(1) name: unnamed; blockState: DEFAULT; state: INPROGRESS, xid/subid/cid: 0/1/0
2021-02-05 14:26:26.015 CET [462715] cbe@postgres DEBUG:  received password packet
2021-02-05 14:26:26.017 CET [462715] cbe@postgres DEBUG:  CommitTransaction(1) name: unnamed; blockState: STARTED; state: INPROGRESS, xid/subid/cid: 0/1/0
2021-02-05 14:26:26.018 CET [462715] cbe@postgres DEBUG:  StartTransaction(1) name: unnamed; blockState: DEFAULT; state: INPROGRESS, xid/subid/cid: 0/1/0
2021-02-05 14:26:26.020 CET [462715] cbe@postgres DEBUG:  CommitTransaction(1) name: unnamed; blockState: STARTED; state: INPROGRESS, xid/subid/cid: 0/1/0
2021-02-05 14:26:26.022 CET [462715] cbe@postgres DEBUG:  StartTransaction(1) name: unnamed; blockState: DEFAULT; state: INPROGRESS, xid/subid/cid: 0/1/0
2021-02-05 14:26:26.982 CET [462715] cbe@postgres ERROR:  java.sql.SQLException: An attempt was made to call a PostgreSQL backend function after an elog(ERROR) had been issued
2021-02-05 14:26:26.982 CET [462715] cbe@postgres STATEMENT:  CREATE EXTENSION pljava
2021-02-05 14:26:26.985 CET [462715] cbe@postgres DEBUG:  shmem_exit(0): 6 on_shmem_exit callbacks to make
2021-02-05 14:26:26.985 CET [462715] cbe@postgres DEBUG:  proc_exit(0): 4 callbacks to make
2021-02-05 14:26:26.985 CET [462715] cbe@postgres DEBUG:  shutting down the Java virtual machine
```

Re ppc64el, yeah I guess we can ignore the problem for now...

**Comment by df7cb** [comment]
There are a few compiler warnings around elog(), not sure if these are related?
```
[INFO] Using compiling/linking rules for Linux
In file included from /usr/include/postgresql/13/server/postgres.h:47,
                 from /srv/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/include/pljava/pljava.h:33,
                 from /srv/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/include/pljava/JNICalls.h:18,
                 from /srv/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/include/pljava/PgObject.h:16,
                 from /srv/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/include/pljava/SPI.h:16,
                 from /srv/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/SPI.c:14:
/srv/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/SPI.c: In function ‘Java_org_postgresql_pljava_internal_SPI__1exec’:
/srv/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/include/pljava/pljava.h:148:4: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  148 |    (uint64)mainThreadId); \
      |    ^
/usr/include/postgresql/13/server/utils/elog.h:128:4: note: in definition of macro ‘ereport_domain’
  128 |    __VA_ARGS__, errfinish(__FILE__, __LINE__, PG_FUNCNAME_MACRO); \
      |    ^~~~~~~~~~~
/usr/include/postgresql/13/server/utils/elog.h:215:2: note: in expansion of macro ‘ereport’
  215 |  ereport(elevel, errmsg_internal(__VA_ARGS__))
      |  ^~~~~~~
/srv/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/include/pljava/pljava.h:147:3: note: in expansion of macro ‘elog’
  147 |   elog(DEBUG2, "Set stack base for thread " UINT64_FORMAT, \
      |   ^~~~
/srv/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/SPI.c:125:3: note: in expansion of macro ‘STACK_BASE_PUSH’
  125 |   STACK_BASE_PUSH(env)
      |   ^~~~~~~~~~~~~~~
```

**Comment by jcflack** [comment]
I see that those warnings around `elog` are something I should fix for `i386`, but probably unrelated to the failure at run time. I might need to ask you for a way to log in to the `i386` build box.

I notice that the `amd64` builds are in `pool/main/p/postgresql-pljava`, but they are not mentioned in `Packages` yet, so `apt` sees nothing to upgrade. Is that just waiting on resolving the build issues for `i386`? Would it be possible to release the `amd64` builds and let `i386` wait on resolving the failure there?

Also, once 1.6.2 builds are released, are they marked in some way to require a deliberate upgrade rather than being pulled in automatically? How does that work? I worry a little about a site suddenly discovering they need to figure out their needed policy entries quickly after an unplanned upgrade.

**Comment by jcflack** [comment]
Also, will older (pre-Java 9) platforms that can't get 1.6.2 be able to get 1.5.7? It does fix a bug in 1.5.6.

**Comment by df7cb** [comment]
> Also, will older (pre-Java 9) platforms that can't get 1.6.2 be able to get 1.5.7? It does fix a bug in 1.5.6.

We don't really have support for maintaining several versions in parallel, and the only affected distribution with 1.5.x is stretch, so I'd rather urge people to upgrade.

**Comment by df7cb** [comment]
> I notice that the `amd64` builds are in `pool/main/p/postgresql-pljava`, but they are not mentioned in `Packages` yet, so `apt` sees nothing to upgrade. Is that just waiting on resolving the build issues for `i386`? Would it be possible to release the `amd64` builds and let `i386` wait on resolving the failure there?

Packages go into the *-pgdg-testing suites first, and get promoted from there. This has happened in the meantime.

> Also, once 1.6.2 builds are released, are they marked in some way to require a deliberate upgrade rather than being pulled in automatically? How does that work? I worry a little about a site suddenly discovering they need to figure out their needed policy entries quickly after an unplanned upgrade.

There is debian/NEWS and apt-listchanges which makes people aware of important changes, I added a news item about this back in February: https://salsa.debian.org/postgresql/postgresql-pljava/-/blob/master/debian/NEWS

**Comment by df7cb** [comment]
The ppc64el issue is still present, but doesn't happen always. Yesterday focal/ppc64el/PG13 failed (instead of 12), but went through when retried. buster+bullseye/ppc64el failed consistently, sometimes on PG12, sometimes on PG13.

https://pgdgbuild.dus.dg-i.net/view/Testsuite/job/postgresql-pljava-autopkgtest/45/
https://pgdgbuild.dus.dg-i.net/view/Testsuite/job/postgresql-pljava-autopkgtest/44/

**Comment by df7cb** [comment]
The issue is still present in 1.6.3, intermittently:

PG11+12+13+14 failing on x86 (32-bit) Debian unstable: https://pgdgbuild.dus.dg-i.net/view/Binaries/job/postgresql-pljava-binaries/123/architecture=i386,distribution=sid/consoleFull#console-section-8
PG12+13 failing on ppc64el Ubuntu focal: https://pgdgbuild.dus.dg-i.net/view/Binaries/job/postgresql-pljava-binaries/123/architecture=ppc64el,distribution=focal/consoleFull#console-section-17 (interestingly, the build-time tests went through fine before it failed the very same tests later in the same log in the install-time tests ("autopkgtest"))

**Comment by jcflack** [comment]
The ppc64 failures still look llvmjit-related; I probably still need to spend time on a ppc64 box to try to track those down.

The x86 logs look to me as if they all failed; 9.6 and 10 just failed differently. It's possible that turns out to be something simpler, like a width mismatch somewhere; there isn't anything in the current CI configuration testing on x86, or 32-bit anything.

**Comment by df7cb** [comment]
Fwiw I've finally disabled the ppc64el and i386 builds on apt.postgresql.org, so only amd64 and arm64 remain, where the pljava tests are stable.

Feel free to close this (and #469) if the other architectures are not in scope.

**Comment by jcflack** [comment]
Closing then until there is demand for one of these arches / access to an environment for testing.
